enricher/javareach/reachable.go

```diff
@@ -0,0 +1,413 @@
+			return nil, fmt.Errorf("directory traversal: %s", classFilepath)
+		}
+
+		if !strings.HasSuffix(classFilepath, ".class") {
+			classFilepath += ".class"
+		}
+
+		if _, err := os.Stat(classFilepath); os.IsNotExist(err) {
+			// Class not found in this directory. Move onto the next classpath.
+			continue
+		}
+
+		classFile, err := os.Open(classFilepath)
+		if err != nil {
+			return nil, err
+		}
+		cf, err := ParseClass(classFile)
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("found class in directory", "class", className, "path", classPath)
+
+		return cf, nil
+	}
+
+	return nil, errors.New("class not found")
+}
+
+// isDynamicCodeLoading returns whether a method and its descriptor represents a
+// call to a dynamic code loading method.
+func isDynamicCodeLoading(method string, descriptor string) bool {
+	// https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/ClassLoader.html#loadClass(java.lang.String)
+	if strings.Contains(method, "loadClass") && strings.HasSuffix(descriptor, "Ljava/lang/Class;") {
+		return true
+	}
+
+	// https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/Class.html#forName(java.lang.String)
+	if strings.Contains(method, "forName") && strings.HasSuffix(descriptor, "Ljava/lang/Class;") {
+		return true
+	}
+
+	return false
+}
+
+// isDependencyInjection returns whether a class provides dependency injection functionality.
+func isDependencyInjection(class string) bool {
+	if strings.HasPrefix(class, "javax/inject") {
+		return true
+	}
+
+	if strings.HasPrefix(class, "org/springframework") {
+		return true
+	}
+
+	if strings.HasPrefix(class, "com/google/inject") {
+		return true
+	}
+
+	if strings.HasPrefix(class, "dagger/") {
+		return true
+	}
+
+	return false
+}
+
+// handleDynamicCode handles the enumeration of class reachability when there is
+// dynamic code loading, taking into account a user specified strategy.
+func (r *ReachabilityEnumerator) handleDynamicCode(q *UniqueQueue[string, *ClassFile], class string, strategy DynamicCodeStrategy) error {
+	if strategy == DontHandleDynamicCode {
+		return nil
+	}
+
+	pkgs, err := r.PackageFinder.Find(class)
+	if err != nil {
+		return err
+	}
+
+	// Assume all classes that belong to the package are reachable.
+	if strategy&AssumeAllClassesReachable > 0 {
+		for _, pkg := range pkgs {
+			classes, err := r.PackageFinder.Classes(pkg)
+			if err != nil {
+				return err
+			}
+
+			for _, class := range classes {
+				if q.Seen(class) {
+					continue
+				}
+				cf, err := r.findClass(r.ClassPaths, class)
+				if err == nil {
+					log.Debug("assuming all package classes are reachable", "class", class, "pkg", pkg)
+					q.Push(class, cf)
+				} else {
+					log.Error("failed to find class", "class", class, "from", pkg, "err", err)
+				}
+			}
+		}
+	}
+	// Assume all classes that belong to the direct dependencies of the package
+	// are reachable.
+	// TODO: implement this.
```

Comment:
All TODOs should have a bug tracking them.

Reply:
As in this won't merge unless there's a bug tracking it, and referenced like so: TODO(b/12345):